<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="dcterms.date" content="2023-11-28">

<title>blog - Breast Cancer Cell Lines Dataset</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../../">
<script src="../../../site_libs/quarto-html/quarto.js"></script>
<script src="../../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script>
<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


<link rel="stylesheet" href="../../../styles.css">
<meta property="og:title" content="blog - Breast Cancer Cell Lines Dataset">
<meta property="og:description" content="">
<meta property="og:image" content="https://avivleemann.github.io/blog/blog/posts/2023-11-28-single_cell/SCRNA-seq-BC_files/figure-html/cell-6-output-1.png">
<meta property="og:site-name" content="blog">
<meta property="og:image:height" content="486">
<meta property="og:image:width" content="510">
<meta name="twitter:title" content="blog - Breast Cancer Cell Lines Dataset">
<meta name="twitter:description" content="">
<meta name="twitter:image" content="https://avivleemann.github.io/blog/blog/posts/2023-11-28-single_cell/SCRNA-seq-BC_files/figure-html/cell-6-output-1.png">
<meta name="twitter:image-height" content="486">
<meta name="twitter:image-width" content="510">
<meta name="twitter:card" content="summary_large_image">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../../../index.html">
    <span class="navbar-title">blog</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../../blog/about.html" rel="" target="">
 <span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../index.html" rel="" target="">
 <span class="menu-text">Posts</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/avivleemann" rel="" target=""><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://LinkedIn.com/in/avivleemann" rel="" target=""><i class="bi bi-linkedin" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-full page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#packages" id="toc-packages" class="nav-link active" data-scroll-target="#packages">Packages:</a></li>
  <li><a href="#data" id="toc-data" class="nav-link" data-scroll-target="#data">Data:</a></li>
  <li><a href="#exploratory-data-analysis" id="toc-exploratory-data-analysis" class="nav-link" data-scroll-target="#exploratory-data-analysis">Exploratory Data Analysis:</a></li>
  <li><a href="#umap-plot" id="toc-umap-plot" class="nav-link" data-scroll-target="#umap-plot">Umap plot</a></li>
  <li><a href="#violine-plot-of-marker-genes" id="toc-violine-plot-of-marker-genes" class="nav-link" data-scroll-target="#violine-plot-of-marker-genes">Violine plot of marker genes</a></li>
  <li><a href="#dotplot-of-marker-genes" id="toc-dotplot-of-marker-genes" class="nav-link" data-scroll-target="#dotplot-of-marker-genes">Dotplot of marker genes</a></li>
  <li><a href="#heatmap-of-clusters-cell-type-and-leiden" id="toc-heatmap-of-clusters-cell-type-and-leiden" class="nav-link" data-scroll-target="#heatmap-of-clusters-cell-type-and-leiden">heatmap of clusters cell type and leiden</a></li>
  <li><a href="#save-data-file" id="toc-save-data-file" class="nav-link" data-scroll-target="#save-data-file">Save data file</a></li>
  <li><a href="#conclusion" id="toc-conclusion" class="nav-link" data-scroll-target="#conclusion">conclusion:</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References:</a></li>
  </ul>
<div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/avivleemann/blog/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content column-page-left" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Breast Cancer Cell Lines Dataset</h1>
  <div class="quarto-categories">
    <div class="quarto-category">SCRNA-seq</div>
    <div class="quarto-category">scanpy</div>
    <div class="quarto-category">Breast cancer</div>
  </div>
  </div>



<div class="quarto-title-meta">

    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">2023-11-28</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->
<section id="packages" class="level2">
<h2 class="anchored" data-anchor-id="packages">Packages:</h2>
<div class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> matplotlib.pyplot <span class="im">import</span> rc_context</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> io</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> anndata <span class="im">as</span> ad</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> leidenalg</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> warnings</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>warnings.filterwarnings(<span class="st">'ignore'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="data" class="level2">
<h2 class="anchored" data-anchor-id="data">Data:</h2>
<div class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> sc.read_h5ad(<span class="st">'GSE173634_Human_BreastCancer_32CellLines_Y2021_35276Cells_Figshare_Counts_Gambardella_diBernardo.h5ad'</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>adata.var_names_make_unique()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>adata</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="3">
<pre><code>AnnData object with n_obs × n_vars = 35276 × 47096
    obs: 'FullCode', 'Cell Type'
    var: 'Ensembl', 'Entrezgene', 'Type of gene', 'Map location', 'chr', 'start', 'end', 'strand', 'Count Pubmed', 'Count Pathways', 'Name', 'Other names'</code></pre>
</div>
</div>
<div class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">'Cell Type'</span>].unique()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="4">
<pre><code>['AU565', 'HCC1937', 'HCC38', 'MDAMB468', 'EFM19', ..., 'HCC1954', 'KPL1', 'ZR751', 'MX1', 'MDAMB453']
Length: 32
Categories (32, object): ['AU565', 'BT20', 'BT474', 'BT483', ..., 'MDAMB468', 'MX1', 'T47D', 'ZR751']</code></pre>
</div>
</div>
</section>
<section id="exploratory-data-analysis" class="level2">
<h2 class="anchored" data-anchor-id="exploratory-data-analysis">Exploratory Data Analysis:</h2>
<p>Show those genes that yield the highest fraction of counts in each single cell, across all cells.</p>
<div class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>sc.pl.highest_expr_genes(adata, n_top<span class="op">=</span><span class="dv">20</span>,)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="SCRNA-seq-BC_files/figure-html/cell-6-output-1.png" class="img-fluid"></p>
</div>
</div>
<p>Basic filtering of cells and genes.</p>
<p>Annotate the group of mitochondrial genes as ‘mt’.</p>
<div class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>sc.pp.filter_cells(adata, min_genes<span class="op">=</span><span class="dv">200</span>)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>sc.pp.filter_genes(adata, min_cells<span class="op">=</span><span class="dv">3</span>)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>adata.var[<span class="st">'mt'</span>] <span class="op">=</span> adata.var_names.<span class="bu">str</span>.startswith(<span class="st">'MT-'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>With pp.calculate_qc_metrics, we can compute many metrics very efficiently</p>
<div class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>sc.pp.calculate_qc_metrics(adata, qc_vars<span class="op">=</span>[<span class="st">'mt'</span>], percent_top<span class="op">=</span><span class="va">None</span>, log1p<span class="op">=</span><span class="va">False</span>, inplace<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>sc.pl.violin(adata, [<span class="st">'n_genes_by_counts'</span>, <span class="st">'total_counts'</span>, <span class="st">'pct_counts_mt'</span>],jitter<span class="op">=</span><span class="fl">0.4</span>, multi_panel<span class="op">=</span><span class="va">True</span>, save<span class="op">=</span><span class="st">'_QC.png'</span>,)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>WARNING: saving figure to file figures/violin_QC.png</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="SCRNA-seq-BC_files/figure-html/cell-8-output-2.png" class="img-fluid"></p>
</div>
</div>
<p>A violin plot of some of the computed quality measures:</p>
<ul>
<li><p>the number of genes expressed in the count matrix.</p></li>
<li><p>the total counts per cell.</p></li>
<li><p>the percentage of counts in mitochondrial genes.</p></li>
</ul>
<div class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>sc.pl.scatter(adata, x<span class="op">=</span><span class="st">'total_counts'</span>, y<span class="op">=</span><span class="st">'pct_counts_mt'</span>, save<span class="op">=</span><span class="st">'_QC_total_pct_mt.png'</span>)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>sc.pl.scatter(adata, x<span class="op">=</span><span class="st">'total_counts'</span>, y<span class="op">=</span><span class="st">'n_genes_by_counts'</span>, save<span class="op">=</span><span class="st">'_QC_total_n_genes.png'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>WARNING: saving figure to file figures/scatter_QC_total_pct_mt.png
WARNING: saving figure to file figures/scatter_QC_total_n_genes.png</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="SCRNA-seq-BC_files/figure-html/cell-9-output-2.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="SCRNA-seq-BC_files/figure-html/cell-9-output-3.png" class="img-fluid"></p>
</div>
</div>
<p>no metochondrial genes in the dataset.</p>
<p>filtering out n_genes_by_counts &lt; 8000.</p>
<p>Total-count normalized, and Logarithmize the data.</p>
<div class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> adata[adata.obs.n_genes_by_counts <span class="op">&lt;</span> <span class="dv">8000</span>, :]</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>sc.pp.normalize_total(adata, target_sum<span class="op">=</span><span class="fl">1e4</span>)</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>sc.pp.log1p(adata)</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>sc.pp.highly_variable_genes(adata, min_mean<span class="op">=</span><span class="fl">0.0125</span>, max_mean<span class="op">=</span><span class="dv">3</span>, min_disp<span class="op">=</span><span class="fl">0.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>sc.pl.highly_variable_genes(adata, save<span class="op">=</span><span class="st">'_HVG.png'</span>,)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>WARNING: saving figure to file figures/filter_genes_dispersion_HVG.png</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="SCRNA-seq-BC_files/figure-html/cell-11-output-2.png" class="img-fluid"></p>
</div>
</div>
<p>Identify highly-variable genes.</p>
<div class="cell" data-execution_count="11">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>adata.raw <span class="op">=</span> adata</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> adata[:, adata.var.highly_variable]</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>sc.pp.regress_out(adata, [<span class="st">'total_counts'</span>, <span class="st">'pct_counts_mt'</span>])</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>sc.pp.scale(adata, max_value<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>sc.tl.pca(adata, svd_solver<span class="op">=</span><span class="st">'arpack'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Reduce the dimensionality of the data by running principal component analysis (PCA), which reveals the main axes of variation and denoises the data.</p>
<p>a scatter plot in the PCA coordinates color=‘Cell Type’.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># rc_context is used for the figure size,</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> rc_context({<span class="st">'figure.figsize'</span>: (<span class="dv">6</span>, <span class="dv">6</span>)}):</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>    sc.pl.pca(adata, color<span class="op">=</span><span class="st">'Cell Type'</span>, save<span class="op">=</span><span class="st">'_PCA.png'</span>,)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>WARNING: saving figure to file figures/pca_PCA.png</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="SCRNA-seq-BC_files/figure-html/cell-13-output-2.png" class="img-fluid"></p>
</div>
</div>
<p>the contribution of single PCs to the total variance in the data.</p>
<p>This gives us information about how many PCs we should consider in order to compute the neighborhood relations of cells.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>sc.pl.pca_variance_ratio(adata, log<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="SCRNA-seq-BC_files/figure-html/cell-14-output-1.png" class="img-fluid"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>sc.tl.tsne(adata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The file that will store the analysis results</p>
<p>compute clusters using the leiden method and store the results with the name <code>clusters</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>sc.pp.neighbors(adata, n_neighbors<span class="op">=</span><span class="dv">10</span>, n_pcs<span class="op">=</span><span class="dv">40</span>)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>sc.tl.leiden(adata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="56">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>sc.tl.paga(adata)</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>sc.pl.paga(adata, plot<span class="op">=</span><span class="va">False</span>)  <span class="co"># remove `plot=False` if you want to see the coarse-grained graph</span></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>sc.tl.umap(adata, init_pos<span class="op">=</span><span class="st">'paga'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="umap-plot" class="level2">
<h2 class="anchored" data-anchor-id="umap-plot">Umap plot</h2>
<p>As we set the .raw attribute of adata, the previous plots showed the “raw” (normalized, logarithmized, but uncorrected) gene expression.</p>
<p>You can also plot the scaled and corrected gene expression by explicitly stating that you don’t want to use .raw.</p>
<p>umap plot by clusters leiden and cell type.</p>
<div class="cell" data-execution_count="105">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(adata, color<span class="op">=</span><span class="st">'Cell Type'</span>, use_raw<span class="op">=</span><span class="va">False</span>,save<span class="op">=</span><span class="st">'_UMAP_cell_type.png'</span>,)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>WARNING: saving figure to file figures/umap_UMAP_cell_type.png</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="SCRNA-seq-BC_files/figure-html/cell-18-output-2.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-execution_count="58">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">'Cell Type'</span>].unique()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="58">
<pre><code>['AU565', 'HCC1937', 'HCC38', 'MDAMB468', 'EFM19', ..., 'HCC1954', 'KPL1', 'ZR751', 'MX1', 'MDAMB453']
Length: 32
Categories (32, object): ['AU565', 'BT20', 'BT474', 'BT483', ..., 'MDAMB468', 'MX1', 'T47D', 'ZR751']</code></pre>
</div>
</div>
<p>there is 32 cell types in the dataset, leiden clusters is 36 clusters.</p>
<p>found more leiden clusters than cell types.</p>
<div class="cell" data-execution_count="106">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(adata, color<span class="op">=</span>[<span class="st">'leiden'</span>], use_raw<span class="op">=</span><span class="va">False</span>,save<span class="op">=</span><span class="st">'_UMAP_leiden.png'</span>,)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>WARNING: saving figure to file figures/umap_UMAP_leiden.png</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="SCRNA-seq-BC_files/figure-html/cell-20-output-2.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-execution_count="107">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(adata, color<span class="op">=</span>[<span class="st">'leiden'</span>, <span class="st">'Cell Type'</span>, <span class="st">'PALLD'</span>],save<span class="op">=</span><span class="st">'_UMAP_leiden_cell_type_PALLD.png'</span>,)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>WARNING: saving figure to file figures/umap_UMAP_leiden_cell_type_PALLD.png</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="SCRNA-seq-BC_files/figure-html/cell-21-output-2.png" class="img-fluid"></p>
</div>
</div>
<p>Finding marker genes.</p>
<p>Let us compute a ranking for the highly differential genes in each cluster.</p>
<p>For this, by default, the .raw attribute of AnnData is used in case it has been initialized before.</p>
<p>The simplest and fastest method to do so is the t-test.</p>
<div class="cell" data-execution_count="80">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>sc.tl.rank_genes_groups(adata, <span class="st">'leiden'</span>, method<span class="op">=</span><span class="st">'t-test'</span>)</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="co"># sc.pl.rank_genes_groups(adata, n_genes=25, sharey=False)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The result of a Wilcoxon rank-sum (Mann-Whitney-U) test is very similar.</p>
<p>We recommend using the latter in publications, see e.g., Sonison &amp; Robinson (2018).</p>
<p>You might also consider much more powerful differential testing packages like MAST, limma, DESeq2 and, for python, the recent diffxpy.</p>
<div class="cell" data-execution_count="81">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># sc.tl.rank_genes_groups(adata, 'leiden', method='wilcoxon')</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a><span class="co"># sc.pl.rank_genes_groups(adata, n_genes=25, sharey=False)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Show the 10 top ranked genes per cluster</p>
<div class="cell" data-execution_count="70">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>pd.DataFrame(adata.uns[<span class="st">'rank_genes_groups'</span>][<span class="st">'names'</span>]).head(<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="70">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">0</th>
<th data-quarto-table-cell-role="th">1</th>
<th data-quarto-table-cell-role="th">2</th>
<th data-quarto-table-cell-role="th">3</th>
<th data-quarto-table-cell-role="th">4</th>
<th data-quarto-table-cell-role="th">5</th>
<th data-quarto-table-cell-role="th">6</th>
<th data-quarto-table-cell-role="th">7</th>
<th data-quarto-table-cell-role="th">8</th>
<th data-quarto-table-cell-role="th">9</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">27</th>
<th data-quarto-table-cell-role="th">28</th>
<th data-quarto-table-cell-role="th">29</th>
<th data-quarto-table-cell-role="th">30</th>
<th data-quarto-table-cell-role="th">31</th>
<th data-quarto-table-cell-role="th">32</th>
<th data-quarto-table-cell-role="th">33</th>
<th data-quarto-table-cell-role="th">34</th>
<th data-quarto-table-cell-role="th">35</th>
<th data-quarto-table-cell-role="th">36</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>VIM</td>
<td>MAGEA4</td>
<td>RPL19</td>
<td>MIEN1</td>
<td>RARRES1</td>
<td>SEC61G</td>
<td>S100A2</td>
<td>RPL13</td>
<td>GSTP1</td>
<td>RARRES1</td>
<td>...</td>
<td>ANKRD30A</td>
<td>RPL19</td>
<td>ENSG00000218227</td>
<td>PSMB3</td>
<td>NUCB2</td>
<td>SERPINI1</td>
<td>RAB11FIP1</td>
<td>RPS20</td>
<td>LDHB</td>
<td>RPS28</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>CAV1</td>
<td>SNHG5</td>
<td>PSMB3</td>
<td>S100A9</td>
<td>PDZK1IP1</td>
<td>CCT6A</td>
<td>RPS20</td>
<td>RPL14</td>
<td>LDHB</td>
<td>PDZK1IP1</td>
<td>...</td>
<td>RAB11FIP1</td>
<td>ENSG00000218227</td>
<td>RPL19</td>
<td>RPL19</td>
<td>SYCP2</td>
<td>VDAC1</td>
<td>MTCO1P40</td>
<td>RPS6</td>
<td>GSTP1</td>
<td>RPL37</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>LDHB</td>
<td>NDUFS5</td>
<td>RPL23</td>
<td>S100A8</td>
<td>SLPI</td>
<td>CHCHD2</td>
<td>RPS6</td>
<td>VIM</td>
<td>MDK</td>
<td>SLPI</td>
<td>...</td>
<td>PLPBP</td>
<td>KRT81</td>
<td>PSMD3</td>
<td>MIEN1</td>
<td>ATP5F1E</td>
<td>FGL2</td>
<td>CRABP2</td>
<td>RPL7</td>
<td>EPCAM</td>
<td>RPS10</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>LGALS1</td>
<td>RPL32</td>
<td>MIEN1</td>
<td>NDUFS6</td>
<td>S100A9</td>
<td>EIF3M</td>
<td>TMSB10</td>
<td>RPS16</td>
<td>RPL22L1</td>
<td>S100A9</td>
<td>...</td>
<td>SLC9A3R1</td>
<td>NDUFB9</td>
<td>LDHA</td>
<td>RPL23</td>
<td>AKAP9</td>
<td>SKP1</td>
<td>TOB1</td>
<td>RPL26</td>
<td>TPT1</td>
<td>UBL5</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>RPS6</td>
<td>LDHB</td>
<td>DYNLRB1</td>
<td>S100A7</td>
<td>CRABP2</td>
<td>SAT1</td>
<td>MIR205HG</td>
<td>RPL13A</td>
<td>ITM2B</td>
<td>CRABP2</td>
<td>...</td>
<td>SYCP2</td>
<td>RPL7</td>
<td>HSP90AB3P</td>
<td>DYNLRB1</td>
<td>PABPC1</td>
<td>GOLIM4</td>
<td>KRT19</td>
<td>RPS28</td>
<td>RPL22L1</td>
<td>SSR3</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">5</td>
<td>PFN1</td>
<td>FTL</td>
<td>YWHAB</td>
<td>STARD3</td>
<td>TAGLN2</td>
<td>MYL12B</td>
<td>CSTA</td>
<td>SNHG5</td>
<td>ITM2C</td>
<td>S100A8</td>
<td>...</td>
<td>SLC39A6</td>
<td>MAL2</td>
<td>COTL1</td>
<td>YWHAB</td>
<td>PEBP1</td>
<td>TCF7</td>
<td>GATA3</td>
<td>RPS14</td>
<td>SLC12A2</td>
<td>PAGE2</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">6</td>
<td>HACD1</td>
<td>AKR1B1</td>
<td>PSMD3</td>
<td>KRT15</td>
<td>IFI27</td>
<td>KRT19</td>
<td>RPL7</td>
<td>RPL27A</td>
<td>HMGA2</td>
<td>B2M</td>
<td>...</td>
<td>ANKRD30B</td>
<td>S100P</td>
<td>PSMA7</td>
<td>PSMD3</td>
<td>ALOX15B</td>
<td>RPS15P4</td>
<td>CRIP2</td>
<td>RPL21</td>
<td>TMEM132D-AS1</td>
<td>RPL41</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">7</td>
<td>CTSB</td>
<td>UCHL1</td>
<td>CISD3</td>
<td>MRPL36</td>
<td>B2M</td>
<td>UBL5</td>
<td>RPS28</td>
<td>PTTG1IP</td>
<td>PLS3</td>
<td>IFI27</td>
<td>...</td>
<td>ZNF703</td>
<td>MRPL13</td>
<td>CDK12</td>
<td>CISD3</td>
<td>MUCL1</td>
<td>MGST2</td>
<td>CLU</td>
<td>RPS18</td>
<td>GTSF1</td>
<td>S100A6</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">8</td>
<td>TMSB10</td>
<td>VIM</td>
<td>KRT19</td>
<td>ATP5F1E</td>
<td>S100A8</td>
<td>LANCL2</td>
<td>RPL27</td>
<td>S100A2</td>
<td>TPM1</td>
<td>TAGLN2</td>
<td>...</td>
<td>ERLIN2</td>
<td>EIF1</td>
<td>OXTR</td>
<td>CALM1</td>
<td>GNAS</td>
<td>RPS26</td>
<td>ERLIN2</td>
<td>RPL27</td>
<td>GPX2</td>
<td>GSTP1</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">9</td>
<td>ENO1</td>
<td>RACK1</td>
<td>PFDN4</td>
<td>RACK1</td>
<td>CDKN2A</td>
<td>RPL10</td>
<td>RPL26</td>
<td>KLF6</td>
<td>PEG10</td>
<td>IGFBP3</td>
<td>...</td>
<td>PIK3R1</td>
<td>KRT19</td>
<td>YWHAB</td>
<td>MRPL45</td>
<td>HSPB1</td>
<td>RPS27</td>
<td>PBX1</td>
<td>RPL10A</td>
<td>GAGE1</td>
<td>EEF1A1</td>
</tr>
</tbody>
</table>

<p>10 rows × 37 columns</p>
</div>
</div>
</div>
<p>Get a table with the scores and groups.</p>
<div class="cell" data-execution_count="71">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>result <span class="op">=</span> adata.uns[<span class="st">'rank_genes_groups'</span>]</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>groups <span class="op">=</span> result[<span class="st">'names'</span>].dtype.names</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>pd.DataFrame(</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>    {group <span class="op">+</span> <span class="st">'_'</span> <span class="op">+</span> key[:<span class="dv">1</span>]: result[key][group]</span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> group <span class="kw">in</span> groups <span class="cf">for</span> key <span class="kw">in</span> [<span class="st">'names'</span>, <span class="st">'pvals'</span>]}).head(<span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="71">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">0_n</th>
<th data-quarto-table-cell-role="th">0_p</th>
<th data-quarto-table-cell-role="th">1_n</th>
<th data-quarto-table-cell-role="th">1_p</th>
<th data-quarto-table-cell-role="th">2_n</th>
<th data-quarto-table-cell-role="th">2_p</th>
<th data-quarto-table-cell-role="th">3_n</th>
<th data-quarto-table-cell-role="th">3_p</th>
<th data-quarto-table-cell-role="th">4_n</th>
<th data-quarto-table-cell-role="th">4_p</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">32_n</th>
<th data-quarto-table-cell-role="th">32_p</th>
<th data-quarto-table-cell-role="th">33_n</th>
<th data-quarto-table-cell-role="th">33_p</th>
<th data-quarto-table-cell-role="th">34_n</th>
<th data-quarto-table-cell-role="th">34_p</th>
<th data-quarto-table-cell-role="th">35_n</th>
<th data-quarto-table-cell-role="th">35_p</th>
<th data-quarto-table-cell-role="th">36_n</th>
<th data-quarto-table-cell-role="th">36_p</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>VIM</td>
<td>0.0</td>
<td>MAGEA4</td>
<td>0.0</td>
<td>RPL19</td>
<td>0.0</td>
<td>MIEN1</td>
<td>0.0</td>
<td>RARRES1</td>
<td>0.0</td>
<td>...</td>
<td>SERPINI1</td>
<td>0.000000e+00</td>
<td>RAB11FIP1</td>
<td>1.929064e-285</td>
<td>RPS20</td>
<td>1.099089e-263</td>
<td>LDHB</td>
<td>8.923028e-91</td>
<td>RPS28</td>
<td>2.214358e-88</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>CAV1</td>
<td>0.0</td>
<td>SNHG5</td>
<td>0.0</td>
<td>PSMB3</td>
<td>0.0</td>
<td>S100A9</td>
<td>0.0</td>
<td>PDZK1IP1</td>
<td>0.0</td>
<td>...</td>
<td>VDAC1</td>
<td>0.000000e+00</td>
<td>MTCO1P40</td>
<td>2.360296e-266</td>
<td>RPS6</td>
<td>6.059649e-228</td>
<td>GSTP1</td>
<td>1.702707e-76</td>
<td>RPL37</td>
<td>2.238960e-84</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>LDHB</td>
<td>0.0</td>
<td>NDUFS5</td>
<td>0.0</td>
<td>RPL23</td>
<td>0.0</td>
<td>S100A8</td>
<td>0.0</td>
<td>SLPI</td>
<td>0.0</td>
<td>...</td>
<td>FGL2</td>
<td>8.300303e-321</td>
<td>CRABP2</td>
<td>1.553926e-286</td>
<td>RPL7</td>
<td>2.985755e-228</td>
<td>EPCAM</td>
<td>4.671848e-74</td>
<td>RPS10</td>
<td>6.693325e-72</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>LGALS1</td>
<td>0.0</td>
<td>RPL32</td>
<td>0.0</td>
<td>MIEN1</td>
<td>0.0</td>
<td>NDUFS6</td>
<td>0.0</td>
<td>S100A9</td>
<td>0.0</td>
<td>...</td>
<td>SKP1</td>
<td>0.000000e+00</td>
<td>TOB1</td>
<td>3.832407e-222</td>
<td>RPL26</td>
<td>8.334210e-234</td>
<td>TPT1</td>
<td>8.243748e-70</td>
<td>UBL5</td>
<td>3.045124e-70</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>RPS6</td>
<td>0.0</td>
<td>LDHB</td>
<td>0.0</td>
<td>DYNLRB1</td>
<td>0.0</td>
<td>S100A7</td>
<td>0.0</td>
<td>CRABP2</td>
<td>0.0</td>
<td>...</td>
<td>GOLIM4</td>
<td>3.952525e-323</td>
<td>KRT19</td>
<td>1.607189e-229</td>
<td>RPS28</td>
<td>6.225240e-219</td>
<td>RPL22L1</td>
<td>1.388007e-67</td>
<td>SSR3</td>
<td>2.703118e-66</td>
</tr>
</tbody>
</table>

<p>5 rows × 74 columns</p>
</div>
</div>
</div>
</section>
<section id="violine-plot-of-marker-genes" class="level2">
<h2 class="anchored" data-anchor-id="violine-plot-of-marker-genes">Violine plot of marker genes</h2>
<p>If we want a more detailed view for a certain group, use sc.pl.rank_genes_groups_violin.</p>
<div class="cell" data-execution_count="108">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>sc.pl.rank_genes_groups_violin(adata, groups<span class="op">=</span><span class="st">'0'</span>, n_genes<span class="op">=</span><span class="dv">8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="SCRNA-seq-BC_files/figure-html/cell-26-output-1.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="dotplot-of-marker-genes" class="level2">
<h2 class="anchored" data-anchor-id="dotplot-of-marker-genes">Dotplot of marker genes</h2>
<p>plot specific gene in dotplot by cell type.</p>
<div class="cell" data-execution_count="109">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>sc.pl.dotplot(adata, [<span class="st">'PALLD'</span>,<span class="st">'VIM'</span>, <span class="st">'CAV1'</span>, <span class="st">'LDHB'</span>, <span class="st">'RPS6'</span>, <span class="st">'HACD1'</span>, <span class="st">'CTSB'</span>,], groupby<span class="op">=</span><span class="st">'Cell Type'</span>,save<span class="op">=</span><span class="st">'_dotplot.png'</span>,)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>WARNING: saving figure to file figures/dotplot__dotplot.png</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="SCRNA-seq-BC_files/figure-html/cell-27-output-2.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="heatmap-of-clusters-cell-type-and-leiden" class="level2">
<h2 class="anchored" data-anchor-id="heatmap-of-clusters-cell-type-and-leiden">heatmap of clusters cell type and leiden</h2>
<div class="cell" data-execution_count="103">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>sc.tl.rank_genes_groups(adata, <span class="st">"Cell Type"</span>, method<span class="op">=</span><span class="st">"t-test"</span>)</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>sc.pl.rank_genes_groups_heatmap(adata, n_genes<span class="op">=</span><span class="dv">10</span>, groupby<span class="op">=</span><span class="st">"Cell Type"</span>,save<span class="op">=</span><span class="st">'_heatmap_cell_type.png'</span>,)</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>sc.tl.rank_genes_groups(adata, <span class="st">"leiden"</span>, method<span class="op">=</span><span class="st">"t-test"</span>)</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>sc.pl.rank_genes_groups_heatmap(adata, n_genes<span class="op">=</span><span class="dv">10</span>, groupby<span class="op">=</span><span class="st">"leiden"</span>,save<span class="op">=</span><span class="st">'_heatmap_leiden.png'</span>,)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>WARNING: Gene labels are not shown when more than 50 genes are visualized. To show gene labels set `show_gene_labels=True`
WARNING: saving figure to file figures/heatmap_heatmap_cell_type.png
WARNING: Gene labels are not shown when more than 50 genes are visualized. To show gene labels set `show_gene_labels=True`
WARNING: saving figure to file figures/heatmap_heatmap_leiden.png</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="SCRNA-seq-BC_files/figure-html/cell-28-output-2.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="SCRNA-seq-BC_files/figure-html/cell-28-output-3.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="save-data-file" class="level2">
<h2 class="anchored" data-anchor-id="save-data-file">Save data file</h2>
<div class="cell" data-execution_count="75">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="co"># results_file = 'GSE173634_Human_BreastCancer_32CellLines_Y2021_35276Cells_Figshare_Counts_Gambardella_diBernardo_RES.h5ad'  </span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a><span class="co"># adata.write(results_file)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="conclusion" class="level2">
<h2 class="anchored" data-anchor-id="conclusion">conclusion:</h2>
<p>The dataset contains 32 cell types, and leiden clusters is 36 clusters.</p>
<p>found more leiden clusters than cell types.</p>
<p>The dataset contains 2700 cells and 33694 genes.</p>
<p>during the analysis, we found that the dataset contains no metochondrial genes.</p>
<p>umap plot by clusters cell type gave us a good separation between the clusters.</p>
<p>PALLD is a gene found to be envolved in the cell migration process and breast cancer metastasis.</p>
<p>in this dataset PALLD has diffrent levels of expression in all the cell types.</p>
<div class="cell" data-execution_count="111">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(adata, color<span class="op">=</span>[ <span class="st">'PALLD'</span>],save<span class="op">=</span><span class="st">'_UMAP_PALLD.png'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>WARNING: saving figure to file figures/umap_UMAP_PALLD.png</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="SCRNA-seq-BC_files/figure-html/cell-30-output-2.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="references" class="level1">
<h1>References:</h1>
<ol type="1">
<li><a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE173634">GSE173634</a></li>
<li><a href="https://www.kaggle.com/datasets/alexandervc/scrnaseq-breast-cancer-cell-lines-atlas">kaggle dataset</a></li>
<li><a href="https://www.nature.com/articles/s41467-022-29358-6">A single-cell analysis of breast cancer cell lines to study tumour heterogeneity and drug response</a></li>
</ol>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>